server/schedule/placement/rule_manager.go

```diff
@@ -296,31 +338,88 @@ func (m *RuleManager) SetRules(rules []*Rule) error {
 	oldRules := make(map[[2]string]*Rule)
 
 	for _, rule := range rules {
-		oldRules[rule.Key()] = m.swapRule(rule)
+		m.addRule(rule, oldRules)
 	}
 
-	ruleList, err := buildRuleList(m.rules)
-	if err == nil {
-		for _, rule := range rules {
-			err = m.store.SaveRule(rule.StoreKey(), rule)
-			if err != nil {
-				break
+	if err := m.tryBuildSave(oldRules); err != nil {
+		return err
+	}
+
+	log.Info("placement rule updated", zap.String("rules", fmt.Sprint(rules)))
+	return nil
+}
+
+func (m *RuleManager) emptyRule(group, id string) ([2]string, *Rule) {
+	key := [2]string{group, id}
+	old, ok := m.rules[key]
+	if ok {
+		delete(m.rules, key)
+	}
+	return key, old
+}
+
+func (m *RuleManager) delRule(t *Batch, oldRules map[[2]string]*Rule) {
+	if !t.MatchID {
+		k, r := m.emptyRule(t.GroupID, t.ID)
+		oldRules[k] = r
+	} else {
+		for key := range m.rules {
+			if key[0] == t.GroupID && strings.HasPrefix(key[1], t.ID) {
+				k, r := m.emptyRule(key[0], key[1])
+				oldRules[k] = r
 			}
 		}
 	}
+}
 
-	if err != nil {
-		for key, rule := range oldRules {
-			if rule == nil {
-				delete(m.rules, key)
-			} else {
-				m.rules[key] = rule
+// BatchAction indicates the operation type
+type BatchAction byte
+
+const (
+	// BatchAdd a placement rule, only need to specify the field *Rule
+	BatchAdd BatchAction = iota + 1
+	// BatchDel a placement rule, only need to specify the field `GroupID`, `ID`, `MatchID`
+	BatchDel
+)
+
+// Batch is for batching placement rule actions. The action type is
+// distinguished by the field `Action`.
+type Batch struct {
+	*Rule               // information of the placement rule to add/delete
+	Action  BatchAction `json:"action"`   // the operation type
+	MatchID bool        `json:"match_id"` // only take effect if action is deletion, will match the prefix of id
```

Comment:
IMO it is better to strict match id by default. So I'm proposing to change it to something like `DeleteByPrefix`.

Reply:
I see. It is 'match id' by default. This name is confusing. Fixed.